src/Agent.Plugins/PipelineArtifact/FileShareProvider.cs

```diff
@@ -0,0 +1,104 @@
+        private const int defaultParallelCount = 1;
+
+        public FileShareProvider(AgentTaskPluginExecutionContext context, CallbackAppTraceSource tracer)
+        {
+            this.context = context;
+            this.tracer = tracer;
+        }
+
+        public async Task DownloadSingleArtifactAsync(PipelineArtifactDownloadParameters downloadParameters, BuildArtifact buildArtifact, CancellationToken cancellationToken)
+        {
+            var downloadRootPath = buildArtifact.Resource.Data + Path.DirectorySeparatorChar + buildArtifact.Name;
+            await this.CopyFileShare(downloadParameters.ProjectId, downloadRootPath, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public async Task DownloadMultipleArtifactsAsync(PipelineArtifactDownloadParameters downloadParameters, IEnumerable<BuildArtifact> buildArtifacts, CancellationToken cancellationToken)
+        {
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var dirPath = Path.Combine(downloadParameters.TargetDirectory, buildArtifact.Name);
+                var downloadRootPath = buildArtifact.Resource.Data;
+                await this.CopyFileShare(downloadParameters.ProjectId, downloadRootPath, dirPath, downloadParameters.MinimatchFilters, cancellationToken, isSingleArtifactDownload: false);
+            }
+        }
+
+        public async Task PublishArtifactAsync(string sourcePath, string destPath, int parallelCount)
+        {
+            await this.DirectoryCopyWithMiniMatch(sourcePath, destPath, CancellationToken.None, parallelCount);
+        }
+
+        private async Task CopyFileShare(Guid projectId, string downloadRootPath, string destPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken, bool isSingleArtifactDownload = true)
+        {
+            //minimatchPatterns = minimatchPatterns.Select(pattern => Path.Combine(downloadRootPath, pattern));
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, this.tracer);
+            await DirectoryCopyWithMiniMatch(downloadRootPath, destPath, cancellationToken, defaultParallelCount, minimatcherFuncs);
+        }
+
+        private async Task DirectoryCopyWithMiniMatch(string sourcePath, string destPath, CancellationToken cancellationToken, int parallelCount = defaultParallelCount, IEnumerable<Func<string, bool>> minimatchFuncs = null)
+        {
+            // If the source path is a file, the system should copy the file to the dest directory directly. 
+            if(File.Exists(sourcePath)) {
+                destPath = destPath + Path.DirectorySeparatorChar + Path.GetFileName(sourcePath);
+                this.context.Output(StringUtil.Loc("CopyFileToDestination", sourcePath, destPath));
+                File.Copy(sourcePath, destPath, true);
+                await Task.CompletedTask;
```

Comment:
I think we meant to return here no? If the source path is of a file, the if statement will complete and will go to line 69 where the following line is:
```
 // Get the subdirectories for the specified directory.
            DirectoryInfo dir = new DirectoryInfo(sourcePath);
```
Since source path is of a file, I'm not sure what the behavior is by calling DirectoryInfo on a file path but I think that's not intended and would lead to unintended consequences.

Reply:
Also one more thing, but if the sourcePath is of a file and the minimatch patterns filters it out, we wouldn't want to download that file right? This logic flow doesn't go through the minimatch filters when the sourcePath is of a single file.